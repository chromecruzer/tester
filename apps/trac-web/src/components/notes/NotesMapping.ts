import _ from "lodash";
import {AuditItemRecord, AuditNoteRecord} from "@trac/datatypes";
type NotesMap = {[id: string]: AuditNoteRecord[]};
export default class NotesMapping {
  private notesMap: NotesMap;
  private itemMap: { [id: string]: AuditItemRecord };
  private auditNotes = [] as AuditNoteRecord[];
  constructor(audit) {
    this.notesMap = _.reduce(audit.notes, (accum, noteRecord: AuditNoteRecord) => {
      if(noteRecord.annotate_type === 'Audit') {
        this.auditNotes.push(noteRecord);
        return accum;
      }
      let entry = accum[noteRecord.annotated_uuid as string];
      if(!_.isArray(entry)) {
        accum[noteRecord.annotated_uuid as string] = [];
        entry = accum[noteRecord.annotated_uuid as string];
      }
      entry.push(noteRecord);
      return accum;
    }, {} as NotesMap );
    this.itemMap = _.reduce(audit.items, (accum, item) => {
      accum[item.uuid] = item;
      return accum;
    }, {})
  }
  getNotes(uuid) {
    return this.notesMap[uuid];
  }
  getItem(uuid) {
    return this.itemMap[uuid];
  }
  getFlattenedAuditNotes(auditHeader) {
    return _.map(this.auditNotes, an => {
      const result = {};
      result[auditHeader] = this.flattenNote(an);
      return result;
    })
  }
  extractItemsWithNotes() {
    return _.map(this.itemMap, (v, k) => {
      const notes = _.reduce(this.getNotes(v.uuid), (accum, n) => {
        accum += `${this.flattenNote(n)}
        `;
        return accum;
      },'');
      return {...v, notes}
    })
  }
  private flattenNote(note) {
    return `${note.content} by ${note.author} on ${note.date_created}`;
  }
}
