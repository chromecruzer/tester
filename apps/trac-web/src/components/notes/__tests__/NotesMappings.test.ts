import NotesMapping from "../NotesMapping";

describe('NotesMapping', () => {
  let notes, items, mappings;
  beforeEach(() => {
    notes = [
      {
        annotate_type: 'Audit',
        annotated_uuid: 'audit1'
      },
      {
        annotate_type: 'Audit',
        annotated_uuid: 'audit2'
      },
      {
        annotate_type: 'Item',
        annotated_uuid: 'item'
      },
      {
        annotate_type: 'Match',
        annotated_uuid: 'item'
      },
      {
        annotate_type: 'Item',
        annotated_uuid: 'other item'
      },
    ];
    items = [
      {
        uuid: 'item',
        lot: 'this is an item'
      },
      {
        uuid: 'other item',
        lot: 'this the other item'
      }
    ];
    mappings = new NotesMapping({location: 'timbuktoo', notes, items});
  });
  it('should fill out maps correctly', () => {
    expect(mappings.itemMap).toMatchSnapshot();
    expect(mappings.notesMap).toMatchSnapshot();
  });
  describe('getNotes', () => {
    it('should get notes for a specified item', () => {
      let actual = mappings.getNotes('other item');
      expect(actual).toMatchSnapshot();
      actual = mappings.getNotes('item');
      expect(actual).toMatchSnapshot();
    });
  });
  describe('getItem', () => {
    it('should get a specified item', () => {
      const actual = mappings.getItem('other item');
      expect(actual).toMatchSnapshot();
    });
  });
});
