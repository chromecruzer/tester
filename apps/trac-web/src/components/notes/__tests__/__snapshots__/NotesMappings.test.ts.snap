// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotesMapping getItem should get a specified item 1`] = `
Object {
  "lot": "this the other item",
  "uuid": "other item",
}
`;

exports[`NotesMapping getNotes should get notes for a specified item 1`] = `
Array [
  Object {
    "annotate_type": "Item",
    "annotated_uuid": "other item",
  },
]
`;

exports[`NotesMapping getNotes should get notes for a specified item 2`] = `
Array [
  Object {
    "annotate_type": "Item",
    "annotated_uuid": "item",
  },
  Object {
    "annotate_type": "Match",
    "annotated_uuid": "item",
  },
]
`;

exports[`NotesMapping should fill out maps correctly 1`] = `
Object {
  "item": Object {
    "lot": "this is an item",
    "uuid": "item",
  },
  "other item": Object {
    "lot": "this the other item",
    "uuid": "other item",
  },
}
`;

exports[`NotesMapping should fill out maps correctly 2`] = `
Object {
  "item": Array [
    Object {
      "annotate_type": "Item",
      "annotated_uuid": "item",
    },
    Object {
      "annotate_type": "Match",
      "annotated_uuid": "item",
    },
  ],
  "other item": Array [
    Object {
      "annotate_type": "Item",
      "annotated_uuid": "other item",
    },
  ],
}
`;
